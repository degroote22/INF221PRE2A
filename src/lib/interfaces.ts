export interface Cost {
  forward: number;
  reverse: number;
  index: number;
}
