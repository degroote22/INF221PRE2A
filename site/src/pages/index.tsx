import * as React from 'react'
import Link from 'gatsby-link'

const IndexPage = ({ data }: { data: any }) => (
  <div>
    <h1>Hi people</h1>
    <p>Welcome to your new Gatsby site.</p>
    <p>Now go build something great!</p>
    {data.allMarkdownRemark.edges.map(({ node }: { node: any }) => (
      <div key={node.id}>
        <p dangerouslySetInnerHTML={{ __html: node.html }} />
      </div>
    ))}
  </div>
)

export const query = graphql`
  query IndexQuery {
    allMarkdownRemark {
      totalCount
      edges {
        node {
          id
          excerpt
          html
        }
      }
    }
  }
`

export default IndexPage
