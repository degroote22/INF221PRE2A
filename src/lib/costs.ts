import { Cost } from "./interfaces";
export const getCost = (points: number[]): number => {
  // O custo é quantos metros a pessoa tem que subir.

  // Constroi pares do caminho

  const pairs = points.reduce(
    (prev, curr, index) => {
      const me = curr;
      const next = points[index + 1];

      // Se for o último, nao vai ter o próximo
      if (!next) {
        return prev;
      }

      const newPair = [me, next] as [number, number];

      return [...prev, newPair];
    },
    [] as [number, number][]
  );

  const cost = pairs.reduce((prev, curr) => {
    const isRising = curr[0] < curr[1];
    const currCost = curr[1] - curr[0];

    if (isRising) {
      return prev + currCost;
    }

    return prev;
  }, 0);

  return cost;
};

export const getCosts = (tracks: string[]): Cost[] => {
  return tracks.map((track, index) => {
    // Picota a string e transforma numa lista de números.
    const ints = track.split(" ").map(i => parseInt(i, 10));

    const [numberOfPoints, ...points] = ints;
    if (numberOfPoints !== points.length) {
      throw Error(
        "Dados inválidos. O tamanho do caminho não bate com o informado."
      );
    }

    const forward = getCost(points);
    const reverse = getCost(points.reverse());
    return {
      index,
      forward,
      reverse
    };
  });
};
