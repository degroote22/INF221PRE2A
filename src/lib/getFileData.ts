import * as fs from "fs";

export const getFileData = (
  filePath: string
): Promise<string> => {
  const error =
    "Não foi possível abrir o arquivo de dados. \
Confira se o arquvio existe e se caminho informado está correto.";
  if (!filePath) {
    return Promise.reject(error);
  }
  // Tornar a operação uma promessa para o código main ficar mais simples.
  return new Promise((resolve, reject) => {
    fs.readFile(filePath, "utf8", (err, data) => {
      if (err) {
        reject(error);
      } else {
        resolve(data);
      }
    });
  });
};

/*
```mermaid

graph TD
    1[1..9]
    2[10-11]
    3[12...15]
    4[16-17]
    5[18-19]

    1 -->|Tipo de dados inválido ou string vazia| 2
    2 --> erro
    1 --> 3
    3 --> 4
    3 -->|Arquivo não encontrado| 5

    4 --> ok
    5 --> erro

    ok --> retorna
    erro --> retorna
```
*/
