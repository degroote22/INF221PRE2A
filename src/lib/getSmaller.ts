import { Cost } from "./interfaces";

export const getSmaller = (costs: Cost[]): number => {
  const smaller = costs.reduce(
    (prev, curr) => {
      // Se é o primeiro ou só tem 1, retorna ele
      if (!prev) {
        return curr;
      }

      const currSmaller = Math.min(
        curr.forward,
        curr.reverse
      );

      const prevSmaller = Math.min(
        prev.forward,
        prev.reverse
      );

      // Só manda o atual se ele for estritamente menor.
      // Isso garante o retorno do menor índice caso sejam iguais.
      if (currSmaller < prevSmaller) {
        return curr;
      }

      // Se o atual não for estritamente menor,
      // o anterior é menor ou igual. Garante que o de
      // menor índice é retornado.
      return prev;
    },
    // Inicializado em null mas com o tipo Cost, que é o retorno.
    // Se o resultado for null, lança um erro
    null as Cost | null
  );

  if (smaller === null) {
    throw Error(
      "Não foi encontrado o menor valor, provavelmente não foi \
informado valor algum na listas de valores desta trilha."
    );
  }
  return smaller.index;
};
