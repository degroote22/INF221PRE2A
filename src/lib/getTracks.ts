export const getTracks = (data: string) => {
  // Separa as linhas por newline
  const lines = data
    .split(/\r?\n/) //unix, windows e mac
    .filter(x => typeof x === "string" && x !== "");

  // A primeira linha é a quantidade e o resto são as linhas.
  // Se os valores são incosistentes, reportar o erro.
  const [linesNumber, ...rest] = lines;
  if (parseInt(linesNumber, 10) !== lines.length - 1) {
    throw Error(
      "Numero de linhas inconsistente com o valor informado."
    );
  }
  return rest;
};
