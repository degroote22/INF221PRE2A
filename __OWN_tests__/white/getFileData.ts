import { getFileData } from "../../src/lib";

describe("Caixa Branca - getFileData", () => {
  test("Retorna os dados certos se passa um caminho válido", () => {
    expect.assertions(1);
    return expect(
      getFileData("./samples/1.txt")
    ).resolves.toMatchSnapshot();
  });

  test("Mostra uma mensagem de erro humana se o argumento é inválido", () => {
    expect.assertions(1);
    return expect(
      getFileData(undefined as any)
    ).rejects.toBe(
      "Não foi possível abrir o arquivo de dados. Confira se o arquvio existe e se caminho informado está correto."
    );
  });

  test("Mostra uma mensagem de erro humana se o arquivo nao existe", () => {
    expect.assertions(1);
    return expect(getFileData("nao_existe")).rejects.toBe(
      "Não foi possível abrir o arquivo de dados. Confira se o arquvio existe e se caminho informado está correto."
    );
  });
});
