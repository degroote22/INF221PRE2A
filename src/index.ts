import { processByFilename } from "./lib";

export const getFilenameFromInput = () => {
  const fileName = process.argv[2];

  if (typeof fileName !== "string") {
    throw Error(
      "Argumento para o nome do arquivo invÃ¡lido"
    );
  }
  return fileName;
};

const main = async () => {
  try {
    const smaller = await processByFilename(
      getFilenameFromInput()
    );
    console.log(smaller);
  } catch (err) {
    // Se houver algum erro no programa, mostrar no console apenas.
    console.error(err);
  }
};

// inicializa o programa
main();
