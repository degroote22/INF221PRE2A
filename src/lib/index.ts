import { getFileData } from "./getFileData";
export * from "./getFileData";
import { getTracks } from "./getTracks";
export * from "./getTracks";
import { getCosts } from "./costs";
export * from "./costs";
import { getSmaller } from "./getSmaller";
export * from "./getSmaller";
export * from "./interfaces";

export const processByFilename = async (
  fileName: string
) => {
  const data = await getFileData(fileName);
  const tracks = getTracks(data);
  const costs = getCosts(tracks);

  // A resposta começa do 1, e até agora usa-se index
  // com base 0.
  const smaller = getSmaller(costs) + 1;

  return smaller;
};
