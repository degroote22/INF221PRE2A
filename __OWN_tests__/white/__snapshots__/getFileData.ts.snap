// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Caixa Branca - getFileData Mostra uma mensagem de erro humana 1`] = `"Não foi possível abrir o arquivo de dados. Confira se o arquvio existe e se caminho informado está correto."`;

exports[`Caixa Branca - getFileData Retorna os dados certos se passa um caminho válido 1`] = `
"5
4 498 500 498 498
10 60 60 70 70 70 70 80 90 90 100
5 200 190 180 170 160
2 1000 900
4 20 20 20 20
"
`;
